core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+	range_mode(FILTER_CONSTRAINT_GREATER),
+	negate(false)
+{
+	if (filter_constraint_is_timestamp(type)) {
+		// For time constraint, default to something completely arbitrary:
+		// Dives in the last year up to next week (newly added dives typically are in the future).
+		QDateTime now = QDateTime::currentDateTimeUtc();
+		data.timestamp_range.from = dateTimeToTimestamp(now.addYears(-1));
+		data.timestamp_range.to = dateTimeToTimestamp(now.addDays(7));
+	} else if (filter_constraint_is_string(type)) {
+		data.string_list = new QStringList;
+	} else if (filter_constraint_is_star(type)) {
+		data.numerical_range.from = 0;
+		data.numerical_range.to = 5;
+	} else if (filter_constraint_is_multiple_choice(type)) {
+		// By default select all the possible items
+		data.multiple_choice = ~0LLU;
+	} else if (filter_constraint_has_time_widget(type) && !filter_constraint_has_date_widget(type)) {
+		// This is a time field. For now let's set it arbitrarily to the whole day.
+		data.numerical_range.from = 0;
+		data.numerical_range.to = 24 * 3600 - 60;
+	} else if (type == FILTER_CONSTRAINT_YEAR) {
+		// Find dives in the last five years.
+		int year = QDateTime::currentDateTimeUtc().date().year();
+		data.numerical_range.from = year - 5;
+		data.numerical_range.to = year;
+	} else {
+		// For numerical data let's try to find sensible defaults based on the unit.
+		// Obviously, these are arbitrary and we might want to save them.
+		switch (type_to_unit(type)) {
+		case FILTER_CONSTRAINT_NO_UNIT:
+		default:
+			data.numerical_range.from = 0;
+			data.numerical_range.to = 0;
+			break;
+		case FILTER_CONSTRAINT_LENGTH_UNIT:
+			// Length: 0-20 m
+			data.numerical_range.from = 0 * 1000;
+			data.numerical_range.to = 20 * 1000;
+			break;
+		case FILTER_CONSTRAINT_TEMPERATURE_UNIT:
+			// Temperature: 20-30°C
+			data.numerical_range.from = C_to_mkelvin(20);
+			data.numerical_range.to = C_to_mkelvin(30);
+			break;
+		case FILTER_CONSTRAINT_WEIGHT_UNIT:
+			// Weight: 0-10 kg
+			data.numerical_range.from = 0 * 1000;
+			data.numerical_range.to = 10 * 1000;
+			break;
+		case FILTER_CONSTRAINT_VOLUMETRIC_FLOW_UNIT:
+			// SAC: 0-10 l/min
+			data.numerical_range.from = 0 * 1000;
+			data.numerical_range.to = 10 * 1000;
+			break;
+		}
+	}
+}
+
+filter_constraint::filter_constraint(const filter_constraint &c) :
+	type(c.type),
+	string_mode(c.string_mode),
+	range_mode(c.range_mode),
+	negate(c.negate)
+{
+	if (filter_constraint_is_timestamp(type))
+		data.timestamp_range = c.data.timestamp_range;
+	else if (filter_constraint_is_string(type))
+		data.string_list = new QStringList(*c.data.string_list);
+	else
+		data.numerical_range = c.data.numerical_range;
+}
+
+filter_constraint &filter_constraint::operator=(const filter_constraint &c)
+{
+	if (filter_constraint_is_string(type))
+		delete data.string_list;
+	type = c.type;
+	string_mode = c.string_mode;
+	range_mode = c.range_mode;
+	negate = c.negate;
+	if (filter_constraint_is_timestamp(type))
```

Comment:
looking at this function and the previous function I notice inconsistency in how you initialized these values, and I wonder if the struct should have a `.mode` sub-struct that you could assign / initialize more conveniently...

Reply:
I don't see any inconsistency...? Constructor initializer lists and assignments are fundamentally different things.

The first construct the subobjects, the second overwrite already existing subobjects.

I don't think that introducing a substructure just for these to functions (constructor and assignment operator) would be worth the additional verbosity.

Reply:
ok